design/extensions.md

```diff
@@ -0,0 +1,272 @@
+
+## Installation Flow
+
+* Extension Manager
+  * Starts up pointing at a “source” registry that contains packages (bundles of an extension and its custom controllers and CRDs) published to it
+  * User requests an extension to be installed in the cluster, which includes everything needed to successfully run that extension
+    * `ExtensionRequest`, which specifies either:
+      * package name (`gitlab`) OR
+      * CRD name (`gitlabcluster.gitlab.com/v1alpha1`)
+        * Note: this code path is exactly the same as dependency resolution
+  * Performs dependency resolution that determines all packages/extensions that are required by this extension and all of its dependent extensions
+  * Pulls all necessary extension packages from the registry
+  * Unpacks them and installs all owned/defined CRDs and transforms the unpackaged content into an `Extension` CRD instance that serves as a record of the install
+  * Starts up the custom controller so that it is in the running state
+  * Marks the `ExtensionRequest` status as succeeded
+
+## `ExtensionRequest` CRD
+
+To commence the installation of new functionality into a Crossplane cluster, an instance of the `ExtensionRequest` CRD should be created.
+The EM will be watching for events on this type and it will begin the process of installing an extension during its reconcile loop.
+
+`ExtensionRequests` can be specified by either a package name or by a CRD type.
+When given a CRD type, the controller will query the registry to find out what package owns that CRD and then it will download that package to proceed with the install.
+This gives more flexibility to how extensions are installed and does not require the requestor to know what package a CRD is defined in.
+
+```yaml
+# request to extend Crossplane with the redis package,
+# using a specific version number
+apiVersion: core.crossplane.io/v1alpha1
+kind: ExtensionRequest
+metadata:
+ name: redis
+spec:
+ source: registry.crossplane.io
+ package: redis-v0.1.0
+status:
+ phase: Succeeded
+---
+# request to extend Crossplane with the package that defines/owns,
+# the rediscluster CRD
+apiVersion: core.crossplane.io/v1alpha1
+kind: ExtensionRequest
+metadata:
+ name: redis
+spec:
+ source: registry.crossplane.io
+ crd: redisclusters.cache.crossplane.io/v1alpha1
+status:
+ phase: Installing
+```
+
+## `Extension` CRD
+
+The `Extension` CRD serves as a record of an installed extension (a custom controller and its CRDs).
+These records make it so that a user or system can query Crossplane and learn all of the functionality that has been installed on it as well as their statuses.
+
+Instances of this CRD can be generated from the filesystem based contents of a package.
+This can be thought of as a translation operation, where the file based content is translated into a YAML based version that is stored in the `Extension` CRD.
```

Comment:
Why translate? Could we just use an instance of this CRD as the metadata inside a package?

Reply:
I don't understand this sentiment. As best I can tell the "filesystem format" is a superset of the data that is captured in the `Extension` resource, so we're going to translate one to the other. It feels like there would be less code complexity and less cognitive overhead to make the "filesystem format" a copy of the `Extension` resource along with the CRD YAML and binary blobs that supplement it.